FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,424 @@
+  }
+
+  private suspend fun prepareAndUploadPasswordProtectedMsgsToFES(account: AccountEntity) =
+    withContext(Dispatchers.IO) {
+      val apiRepository = FlowcryptApiRepository()
+      val keysStorage = KeysStorageImpl.getInstance(applicationContext)
+      val accountSecretKeys = PGPSecretKeyRingCollection(keysStorage.getPGPSecretKeyRings())
+
+      var list: List<MessageEntity>
+      var lastMsgUID = 0L
+      while (true) {
+        list = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+          account = account.email,
+          msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+        )
+
+        if (list.isEmpty()) break
+        val msgEntity = list.firstOrNull { it.uid > lastMsgUID } ?: list[0]
+
+        lastMsgUID = msgEntity.uid
+
+        try {
+          if (msgEntity.isEncrypted == true && msgEntity.isPasswordProtected) {
+            //get msg attachments(including forwarded that can be encrypted)
+            val attachments = getAttachments(msgEntity)
+
+            //create MimeMessage from content + attachments
+            val plainMimeMsgWithAttachments = getMimeMessage(account, msgEntity, attachments)
+
+            //get recipients that will be used to create password-encrypted msg
+            val toCandidates = getRecipients(plainMimeMsgWithAttachments, Message.RecipientType.TO)
+            val ccCandidates = getRecipients(plainMimeMsgWithAttachments, Message.RecipientType.CC)
+            val bccCandidates =
+              getRecipients(plainMimeMsgWithAttachments, Message.RecipientType.BCC)
+
+            if (toCandidates.isEmpty() && ccCandidates.isEmpty() && bccCandidates.isEmpty()) {
+              throw IllegalStateException("Wrong password-protected implementation")
+            }
+
+            //start of creating and uploading a password-protected msg to FES
+            val fromAddress = plainMimeMsgWithAttachments.getFromAddress()
+            val domain = EmailUtil.getDomain(fromAddress)
+            val idToken = getGoogleIdToken()
+            val replyToken = fetchReplyToken(apiRepository, domain, idToken)
+            val messageUploadRequest = MessageUploadRequest(
+              associateReplyToken = replyToken,
+              from = fromAddress,
+              to = toCandidates.map { (it as InternetAddress).address },
+              cc = ccCandidates.map { (it as InternetAddress).address },
+              bcc = bccCandidates.map { (it as InternetAddress).address }
+            )
+
+            //prepare bodyWithReplyToken
+            val replyInfo = Base64.getEncoder().encodeToString(
+              GsonBuilder().create().toJson(messageUploadRequest).toByteArray()
+            )
+
+            val infoDiv = genInfoDiv(replyInfo)
+            val originalText = getDecryptedContentFromMessage(
+              mimeMsgWithAttachments = plainMimeMsgWithAttachments,
+              accountSecretKeys = accountSecretKeys,
+              keysStorage = keysStorage
+            )
```

Comment:
This is confusing. `plainMimeMsgWithAttachments` is encrypted after all? That means my suggestion is wrong and we should not be calling it `plain`?

What is this variable, exactly?

It seems like the character of the variable changes as we use it? If that is the case, maybe whenever the change happens, we should also rename it accordingly?

Reply:
We discussed it [here ](https://github.com/FlowCrypt/flowcrypt-android/pull/1629#discussion_r777738064).

It's realy `plainMimeMsgWithAttachments` that contains `message.asc` as the first MIME part.